libcentrifugo/engineredis.go

```diff
@@ -613,28 +625,129 @@ func (e *RedisEngine) runPubSub() {
+	messageJSON []byte
+	historyKey  string
+	opts        *publishOpts
+	err         *chan error
+}
+
+func (pr *pubRequest) done(err error) {
+	if pr.err == nil {
+		return
+	}
+	*(pr.err) <- err
+}
+
+func (pr *pubRequest) result() error {
+	if pr.err == nil {
+		// No waiting, as caller didn't care about response
+		return nil
+	}
+	return <-*(pr.err)
+}
+
+func fillPublishBatch(ch chan *pubRequest, prs *[]*pubRequest) {
+	for len(*prs) < RedisPublishBatchLimit {
+		select {
+		case pr := <-ch:
+			*prs = append(*prs, pr)
+		default:
+			return
+		}
+	}
+}
+
+func (e *RedisEngine) runPublishPipeline() {
+
 	conn := e.pool.Get()
-	defer conn.Close()
+	err := e.pubScript.Load(conn)
+	if err != nil {
+		logger.ERROR.Println(err)
+		conn.Close()
+		// Can not proceed if script has not been loaded - because we use EVALSHA command for
+		// publishing with history.
+		return
+	}
+	conn.Close()
 
-	var err error
-	if opts == nil {
-		// just publish message into channel.
-		_, err = conn.Do("PUBLISH", chID, message)
-	} else {
-		// publish message into channel and add history message.
-		if opts.HistorySize > 0 && opts.HistoryLifetime > 0 {
-			messageJSON, err := json.Marshal(opts.Message)
+	var prs []*pubRequest
+
+	for {
+		pr := <-e.pubCh
+		prs = append(prs, pr)
+		fillPublishBatch(e.pubCh, &prs)
+
+		conn := e.pool.Get()
+
+		for i := range prs {
+			if prs[i].opts != nil && prs[i].opts.HistorySize > 0 && prs[i].opts.HistoryLifetime > 0 {
+				e.pubScript.SendHash(conn, prs[i].historyKey, prs[i].channel, prs[i].message, prs[i].messageJSON, prs[i].opts.HistorySize, prs[i].opts.HistoryLifetime, prs[i].opts.HistoryDropInactive)
+			} else {
+				conn.Send("PUBLISH", prs[i].channel, prs[i].message)
+			}
+		}
+		err := conn.Flush()
+		if err != nil {
+			for i := range prs {
+				prs[i].done(err)
+			}
+			conn.Close()
+			return
+		}
+		var noScriptError bool
+		for i := range prs {
+			_, err := conn.Receive()
 			if err != nil {
-				logger.ERROR.Println(err)
-				return nil
+				// Check for NOSCRIPT error. In normal circumstances this should never happen.
+				// The only possible situation is when Redis scripts were flushed. In this case
+				// we will return from this func and load publish script from scratch.
+				// Redigo does the same check but for single EVALSHA command: see
+				// https://github.com/garyburd/redigo/blob/master/redis/script.go#L64
+				if e, ok := err.(redis.Error); ok && strings.HasPrefix(string(e), "NOSCRIPT ") {
+					noScriptError = true
+				}
 			}
-			_, err = e.pubScript.Do(conn, e.getHistoryKey(chID), chID, message, messageJSON, opts.HistorySize, opts.HistoryLifetime, opts.HistoryDropInactive)
-		} else {
-			_, err = conn.Do("PUBLISH", chID, message)
+			prs[i].done(err)
+		}
+		conn.Close()
+		if noScriptError {
+			// Start this func from the beginning and LOAD missing script.
+			return
 		}
+		prs = nil
+	}
+}
+
+func (e *RedisEngine) publish(chID ChannelID, message []byte, opts *publishOpts) error {
```

Comment:
HOw about this very simple change to engine interface:

 - Make another method `publishAsync` which has same args but which returns `<-error` instead of `error`.
 - `publish` implementation then becomes:
```go
func (e *RedisEngine) publish(chID ChannelID, message []byte, opts *publishOpts) error {
    return <-e.publishAsync(chID, message, opts)
}
```
 - Cases like `broadcastCmd` above are now way more efficient and can just do something like:
```go
errs := make([]<-error, len(channels))
for i, channel := range channels {
    errs[i] = engine.publishAsync(channel, ...)
}
for _, err := range errs {
    if err != nil {
        // decide how to handle partial errors - just log?
    }
}

```